keps/sig-api-machinery/4346-informer-metrics/README.md

```diff
@@ -0,0 +1,935 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The informer metrics is disabled by default. When enable informer metrics, the newly added metrics will increase CPU/MEM usage. 
+
+If the metrics result memory leak, users can disable the informer metrics.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Introduce the informer metrics struct `informerMetrics` and `eventHandlerMetrics`. It is similar to the existing `workqueue` metrics.
+
+```
+type informerMetrics struct {	
+  clock clock.Clock
+
+  // total number of item in store
+  numbernOfStoredItem  GaugeMetric
+  // total number of item in queue
+  numberOfQueuedItem   GaugeMetric
+  
+  // each eventHandler metrics 
+  eventHandlerMetrics map[string]eventHandlerMetrics
+}
+
+type eventHandlerMetrics struct {
+
+  // number of pending data
+  numberOfPendingNotifications GaugeMetric
+
+  // size of RingGrowring data
+  sizeOfRingGrowing GaugeMetric
+
+  // how long processing an item from informer reflector
+  prcoessDuration  HistogramMetric
+        
+}
+
+// MetricsProvider generates various metrics used by the queue.
+type MetricsProvider interface {
+  // the informer name
+	NewStoredItemMetric(name string) GaugeMetric
+  NewQueuedItemMetric(name string) GaugeMetric
+
+  // the eventHandler name
+  NewPendingNotificationsMetric(name string) GaugeMetric
+  NewRingGrowingMetric(name string) GaugeMetric
+  NewPrcoessDurationMetric(name string) HistogramMetric
+}
+
+```
+
+Add prometheus metrics item in subsystem `informer`
+
+| name                         | labels             | description                                                                |
+|------------------------------|--------------------|----------------------------------------------------------------------------|
+| store_item_total             | informer name      | Total number of item in store                                              |
+| queued_item_total            | informer name      | Total number of item in queue                                              |
+| pending_notifications_total  | eventHandler name  | Total number of pending notifications in eventHandler RingGrowing          |
+| ring_growing_capacity        | eventHandler name  | Capacity of eventHandler RingGrowing                                       |
+| event_process_duration       | eventHandler name  | How long in seconds eventHandler processing an item from RingGrowing takes |
+
+
+Add an environment `ENABLE_INFORMER_METRICS` to enable informer metrics, it is disable by default.
```

Comment:
We now have client-go featuregates.  I suggest using those: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/features/known_features.go#L48

Reply:
Replace it with feature gate InformerMetrics